### ISSUE TITLE
Should remove breaks for OHLC, Candlestick

### ISSUE THREAD

**Issue by mckelvin** [plotly/plotly.js#1382]
[The code to reproduce](https://codepen.io/mckelvin/pen/NdJpjx) is based on [simple-candlestick-chart-example](https://plot.ly/javascript/candlestick-charts/#simple-candlestick-chart-example) . The difference is:

**11th, 12th are not existed** in the dataset (simply because they're weekdays/non-trading days), but candlesticks in the chart should be continued, i.e.: the breaks of 11th, 12th should be removed like [what highcharts do](http://www.highcharts.com/stock/demo/candlestick).

This behaviour should be the default one for OHLC and Candlestick charts, otherwise it will become [terrible for real world dataset](https://plot.ly/~mckelvin/5/).

* * *

https://codepen.io/mckelvin/pen/ygwMQz is my workaround by using `tickvals` and `ticktext`. But there's another problem: **nticks doesn't work when tickmode=array** as mentioned in docs, there will be [too many tick labels](https://plot.ly/~mckelvin/3/#plot) if there're too many ticks.

**Comment by etpinard** [comment]
Agreed. But, I don't like your tone. Saying that  plotly.js _should be_ like Highcharts isn't a very good argument. Thanks.

**Comment by mckelvin** [comment]
@etpinard Thanks for your reply. Sorry but you might misunderstand me. I don't mean to argue for the feature as it's very common in OHLC/Candlestick charts and the link is just provided as a demo which contains the feature I expected. Both plotly and highcharts are great softwares but neither of them is a copycat.

**Comment by etpinard** [comment]
@mckelvin no worries. Thanks for posting!

**Comment by jackparmer** [comment]
Current proposal is to either add a new axis type:
`type: 'date-no-we'`

Or a flag for the axis object:
`show_weekend: false`

cc @jackwluo @cpsievert @santosjorge

Anyone out there interested in sponsoring this one? :clinking_glasses: 
https://plot.ly/products/consulting-and-oem/

**Comment by jackwluo** [comment]
Implement is as a 'financial' flag for axis, since different markets have different holidays

**Comment by jackparmer** [comment]
@jackwluo - That's an interesting point. Can you think of any online reference for which markets respect which holidays?

It looks like mpl and ggplot user have also come across this and have their own workarounds:

[ggplot](http://stackoverflow.com/questions/32893176/ggplot2-remove-blank-space-for-weekends-and-holidays-from-x-axis-dates)
[mpl](http://stackoverflow.com/questions/42155916/how-to-remove-weekends-in-matplotlib-candlestick-chart)

**Comment by jackwluo** [comment]
Seems like NYTimes has a comprehensive database; 

http://markets.on.nytimes.com/research/markets/holidays/holidays.asp?display=market&exchange=LSE

Or individual exchange links below

https://www.nyse.com/markets/hours-calendars
https://www.tsx.com/trading/calendars-and-trading-hours/calendar-and-events
http://www.jpx.co.jp/english/corporate/calendar/

**Comment by mckelvin** [comment]
This issue not only happens on **dates**, but also on other resolutions like `1min`, `5min`, `1h` since the trading session in a day might be 08:00~16:00 (not full 24h). So I think the library should not care about the detailed holidays.

SEE ALSO: https://www.tradingview.com

**Comment by santosjorge** [comment]
In my opinion - I think it is an overkill to expect Plotly having to retrieve holidays/trading sessions for all instruments. You will also have to handle scenarios where there is multiple calendars (ie a GBP swap). 

Most data providers already take into account non-trading days/market sessions into account. That means the data has already been cleaned/adjusted.  

The only thing required would be a layout parameter to create an axis that will omit the dates for which no values exist in the timeseries.

**Comment by Feng-Zihao** [comment]
It's not necessary to retrive all holidays. Every stock market, every country ,may have their own holidays.

Just treat the datetime x_Axis as 'string' instead of 'Date'.

**Comment by frogger72** [comment]
Any update on this?  I think a general solution would be an option to ignore missing dates.  Other solutions require avoiding use of dates for the x-axis which is a big handicap for features eg use rangeslider/rangeselector won't work for array.

**Comment by mckelvin** [comment]
@etpinard Make sense but I don't agree with the priority. IMHO, the missing feature is the largest barrier ahead of the financial users since the OHLC chart is almost the basic chart of all other high level indicator charts. If they can't even try  plotly to draw the basic chart, you can neither expect them to use it and pay for consulting.

More and more financial/trading companies start to use python stack for data analysis (at least I know Optiver is using). Plotly works so well on Jupyter. They are generous to pay for what could save time and bring profit. I'd happy to see good match between plotly and trading companies. That's really a good reason to increase the priority. Please reconsider it.

**Comment by frogger72** [comment]
I'd agree.  If the basic price chart is broken, it's just not a viable solution - it's a real impediment to people even trying.

**Comment by etpinard** [comment]
I apologize for being honest. Locking this issue until tempers cool

**Comment by etpinard** [comment]
From @anothar in https://github.com/plotly/plotly.js/issues/2496 

Good day. Thank you for awesome library.
I created this issue only because I can't post solution in [this thread](https://github.com/plotly/plotly.js/issues/1382). There is no need to download data about holidays for chart without breaks.
Here is [elegant solution](https://codepen.io/anothar/pen/MVvmMz) for daily (or any timeframe) candlestick chart without breaks. The only problem is display text on hover. Instead of date it displays index. This can be fixed when [this issue](https://github.com/plotly/plotly.js/issues/2481) will be solved.

**Comment by nite** [comment]
Note that if you have enough points/ticks you can approximate this currently by switching to a categorical xaxis & setting categoryorder like so: 
```
layout={
    xaxis: {
        type: 'category', 
        categoryorder: 'category ascending'
}}
```
https://codepen.io/nite/pen/gBxoVy?editors=0010

Worth noting that the xaxis then doesn't have an even flow of time, ie. each tick will be an even distance apart, regardless of differing durations between them (eg. if you're plotting a timeseries scatterplot or line plot - less of an issue for grouped plots with even time ticks like candlesticks. 

Would be interesting how & where in the code one would be able to submit a patch @chriddyp @etpinard

**Comment by nicolaskruchten** [comment]
One way forward here would be to add an option to `date` axes to not-render day/hour/whatever-long periods with no data points.

**Comment by gzadigo** [comment]
One way is let the x-coordinate be continueous real numbers like 1,2,3,4,5.... while corresponding text be the date type like Jan1, Jan2,(gap here) Jan5, Jan6, Jan7, .... this can make zoom-in-out easy and showing x-coordinate hints in date format

**Comment by frogger72** [comment]
I believe the proposed workarounds (using category or array types) break rangeselector, which requires date. Is there a workaround that keeps rangeselector functionality?

**Comment by dominhtri1995** [comment]
any update on this ?

**Comment by nicolaskruchten** [comment]
Building on my suggestion above, we could implement this as another axis type, similar to the proposal in #3870 (datetzomit ?)

**Comment by gzadigo** [comment]
Looking forward to seeing this done. 
plotly is beautiful!

**Comment by nicolaskruchten** [comment]
Just thinking ahead: the idea of just dropping time-periods with no data will work for OHLC/Candlestick, but will likely not be enough for something like Gantt charts.

**Comment by gzadigo** [comment]
any update on this feature yet? It is killing me :-(

**Comment by kevinheavey** [comment]
JavaScript is outside my wheelhouse but are we looking to incorporate something like [d3fc-discontinuous-scale](https://d3fc.io/api/discontinuous-scale-api.html)? Live demo [here](https://bl.ocks.org/ColinEberhardt/0a5cc7ca6c256dcd6ef1e2e7ffa468d4)

**Comment by amresh1495** [comment]
Setting the xaxis as object type seems the only viable option which is working.

**Comment by frogger72** [comment]
Easiest would be to implement an xaxis object type called "date-ignore-breaks" or "date-ignore-gaps." This would hopefully preserve current 'date' RangeSlider functionality but ignore gaps, which would be a boon to all timeseries applications.

**Comment by pavitrakumar78** [comment]
Is there any update on this? I think a way to ignore breaks in timeseries data would be better than just ignoring weekends because it generalizes a lot of finance-related charts.

**Comment by mofojed** [comment]
Will the proposed solution simply omit any dates that have no data? It might be the case that we do have a non-holiday day with no data to show as a gap. Will there be a way to accomplish that?

(Love the library!)

**Comment by etpinard** [comment]
In `v1.53.0`, users will be able to remove non-working days and non-working hours using the new `axis.breaks` API for `type: 'date'` cartesian axes.

Some examples:

- breaks using pattern on date axis: https://codepen.io/etpinard/pen/PoqRNoL
- breaks using values on date axis: https://codepen.io/etpinard/pen/OJVvNJG

----

I'll leave this issue open as some folks here appear to prefer the "date as categories" approach - which is another planned feature of ours.

**Comment by nicolaskruchten** [comment]
We're actually going to close this issue as "resolved" with the new `*axis.rangebreaks` attribute available for date axes in the upcoming 1.53 release. If anyone has any additional requests in this vein, I'd ask that they please open a new issue after having had a chance to play with the new version :)

I'll also note that this was a [sponsored feature](https://plotly.com/consulting-and-oem/) and thank our anonymous sponsor for their generous support of open-source ❤️ 🙏
